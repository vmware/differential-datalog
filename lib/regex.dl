/* Bindings for the Rust regular expressions library.
 *
 * See regex crate documentation for description of the regex syntax supported
 * by this library.
 */

/*
 * Check if string matches regular expression.
 *
 * Returns true if the regex matches the string given.
 * Returns false if `regex` is not a valid regular expression or if no match was
 * found.
 */
extern function regex_match(regex: string, text: string): bool

/*
 * Returns the leftmost-first match in `text`. If `regex` is not a valid regular
 * expression or no match exists, then `None` is returned.
 */
extern function regex_first_match(regex: string, text: string): Option<string>

/*
 * Returns all successive non-overlapping matches in `text`.
 * If `regex` is not a valid regular expression or no match exists,
 * an empty vector is returned.
 */
extern function regex_all_matches(regex: string, text: string): Vec<string>
