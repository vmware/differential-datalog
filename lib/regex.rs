extern crate regex;

pub fn regex_regex_match(regex: &String, s: &String) -> bool {
    match regex::Regex::new(regex.as_str()) {
        Result::Err(_) => false,
        Result::Ok(re) => re.is_match(&s.as_str()),
    }
}

pub fn regex_regex_first_match(regex: &String, s: &String) -> std_Option<String> {
    match regex::Regex::new(regex.as_str()) {
        Result::Err(_) => std_Option::std_None,
        Result::Ok(re) => std_Option::from(re.find(&s.as_str()).map(|m| m.as_str().to_string())),
    }
}

pub fn regex_regex_all_matches(regex: &String, s: &String) -> std_Vec<String> {
    match regex::Regex::new(regex.as_str()) {
        Result::Err(_) => std_Vec::new(),
        Result::Ok(re) => {
            let v: Vec<_> = re
                .find_iter(&s.as_str())
                .map(|m| m.as_str().to_string())
                .collect();
            std_Vec::from(v)
        }
    }
}
