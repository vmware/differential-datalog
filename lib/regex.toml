[dependencies.regex]
version = "1.1"
