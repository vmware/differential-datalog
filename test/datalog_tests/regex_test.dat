dump regex_test.RegexTestOutput;
