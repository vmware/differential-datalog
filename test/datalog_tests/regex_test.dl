import regex

relation RegexTestInput(
    regex: string,
    text: string)


RegexTestInput([|\d+|], "a111b222c").
RegexTestInput([|\b\w{13}\b|], "I categorically deny having triskaidekaphobia.").
RegexTestInput([|'([^']+)'\s+\((\d{4})\)|], "Not my favorite movie: 'Citizen Kane' (1941).").
RegexTestInput("[", "foo").

output relation RegexTestOutput(
    regex: string,
    text: string,
    match_found: bool,
    match_found_checked: bool,
    first_match: Option<string>,
    all_matches: Vec<string>)

RegexTestOutput(regex,
                text,
                regex_match(regex(regex), text),
                regex_match(result_unwrap_or(regex_checked(regex), regex("")), text),
                regex_first_match(regex(regex), text),
                regex_all_matches(regex(regex), text)) :-
    RegexTestInput(regex, text).
