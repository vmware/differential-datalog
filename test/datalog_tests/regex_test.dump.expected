regex_test.RegexTestOutput{.regex = "'([^']+)'\\s+\\((\\d{4})\\)", .text = "Not my favorite movie: 'Citizen Kane' (1941).", .match_found = true, .first_match = std.Some{.x = "'Citizen Kane' (1941)"}, .all_matches = ["'Citizen Kane' (1941)"]}
regex_test.RegexTestOutput{.regex = "\\b\\w{13}\\b", .text = "I categorically deny having triskaidekaphobia.", .match_found = true, .first_match = std.Some{.x = "categorically"}, .all_matches = ["categorically"]}
regex_test.RegexTestOutput{.regex = "\\d+", .text = "a111b222c", .match_found = true, .first_match = std.Some{.x = "111"}, .all_matches = ["111", "222"]}
